c++/src/capnp/schema-parser.c++

```diff
@@ -102,7 +102,13 @@ public:
 
   kj::Maybe<kj::Array<const byte>> embedRelative(kj::StringPtr embedPath) override {
     KJ_IF_SOME(importedFile, file->import(embedPath)) {
-      return importedFile->readContent().releaseAsBytes();
+      auto content = importedFile->readContent();
+      // Do not try to read an empty file, but return an empty array (kj::none would indicate a
+      // read error).
+      if (!content.size()) {
+        return kj::Array<const byte>(nullptr);
+      }
+      return content.releaseAsBytes();
```

Comment:
I think we should fix `releaseAsBytes()` and `releaseAsChars()` directly, rather than make all the call sites think about this.

The problem appears to be that these functions allocate a new `Array`, calling the constructor where they pass a custom disposer. The constructor takes the disposer by reference, so if the disposer is null, this is UB, since references can't be null.

Probably, `releaseAsBytes()` needs to check for `disposer == nullptr` like:

```
  inline Array<PropagateConst<T, byte>> releaseAsBytes() {
    // Like asBytes() but transfers ownership.
    static_assert(sizeof(T) == sizeof(byte),
        "releaseAsBytes() only possible on arrays with byte-size elements (e.g. chars).");
    if (disposer == nullptr) return nullptr;
    Array<PropagateConst<T, byte>> result(
        reinterpret_cast<PropagateConst<T, byte>*>(ptr), size_, *disposer);
    ptr = nullptr;
    size_ = 0;
    return result;
  }
```

And similarly for `releaseAsChars()`.

Reply:
👍  that approach should be more comprehensive, updated the commit to use that instead